### PR TITLE
Make postfixOps syntax an error (not just a warning) unless the feature is explicitly enabled [ci: last-only]

### PR DESCRIPTION
**Pull request by som-snytt**
Follows up https://github.com/scala/scala/pull/6559 with requiring postfix flag.

This change makes the postfixOps language feature more opt-in, to discourage newbies from exploiting the feature and falling into its snares.

The tide turned against postfix syntax [on the forum](https://contributors.scala-lang.org/t/lets-drop-postfix-operators/1457).

Codebases relying on postfix syntax will want to enable `-language:postfixOps`.

### CODE REVIEW

**Comment by SethTisue** [comment]
needs rebase

**Comment by SethTisue** [comment]
@som-snytt what's next here?

**Comment by som-snytt** [comment]
@SethTisue IIUC @adriaanm has to defer things he'd rather be doing to do some sort of restarr dance. "Dancing with the Restarrs!" Some people would love to get a shot at that.

**Comment by adriaanm** [comment]
When starr becomes M5, we can rebase and merge this for RC1. Right?

**Comment by SethTisue** [comment]
@som-snytt when you wish upon a STARR you wake up and M5 is far behind you ✨

**Comment by adriaanm** [comment]
🕺 🌠

**Comment by SethTisue** [comment]
compilation errors when building partest

**Comment by som-snytt** [comment]
@SethTisue paulp style! I have a few minutes before taking trash to the dump. On the coast of Maine, everyone is their own garbage person. I think that may be the basis for true democracy.

**Comment by som-snytt** [comment]
I keep thinking haven't I already made this edit once before? For the record, there was one line I had to think about how to parse and amend, so it's not like, infix with postfix, what's the problem?

**Comment by SethTisue** [comment]
```
[test] [error]   - library/compile:compileIncremental failed: Error compiling sbt component 'compiler-interface'
```

looks like this needs the special `checkGoldenUnits` dispensation?

**Comment by som-snytt** [comment]
That should be in starr.

Maybe the update includes an sbt change. The test requires the xsbt file be compiled by itself.

**Comment by som-snytt** [comment]
Indeed, debug says
```
xsbt compat compiled with 15 units!!!
```
I probably won't bother checking if the behavior changed; the test for one unit was just for insurance, since the test only checks trailing characters in the path. Oh wait, it was also nice not to check a possibly long list of units for the special name.

I wonder if @dwijnand knows if `sbt` changed how it compiles `xsbt/Compat.scala`. Edit: the last bump was to 0.13.17, months ago.

Looking at the kludge now, I see it sets the language option, but doesn't restore it. Another fix would have been to inject a root import of the language feature, now that I know how root imports work. One could also imagine encapsulating the special knowledge about the path in SourceFile.

**Comment by dwijnand** [comment]
Well there's https://github.com/sbt/zinc/commit/c60ff2371fa953fd8f07b5aec3a6157612796811, but that's not that recent.

**Comment by som-snytt** [comment]
@dwijnand Thanks. My issue is that I got the idea that the `Compat.scala` files is always compiled by itself, so I checked that first for purposes of ignoring postfix in that file. But now 15 files are compiled with it. (I'm about to check which.)

File: src/library/scala/languageFeature.scala [hunk 1]
```diff
@@ -1,13 +1,11 @@
 /*                     __                                               *\
 **     ________ ___   / /  ___     Scala API                            **
-**    / __/ __// _ | / /  / _ |    (c) 2003-2015, LAMP/EPFL             **
+**    / __/ __// _ | / /  / _ |    (c) 2003-2018, LAMP/EPFL             **
```
Comment (som-snytt) [inline src/library/scala/languageFeature.scala]: What could it hurt? he said.

**Comment by som-snytt** [comment]
Under 0.13.16. (The fix is in 0.13.18, https://github.com/sbt/sbt/pull/4132.)
```
[info] 'compiler-interface' not yet compiled for Scala (unknown). Compiling...
Compiling 15 units: List(API.scala, Analyzer.scala, Command.scala, Compat.scala, CompilerInterface.scala, ConsoleInterface.scala, DelegatingReporter.scala, Dependency.scala, ExtractAPI.scala, ExtractUsedNames.scala, GlobalHelpers.scala, LocateClassFile.scala, Log.scala, Message.scala, ScaladocInterface.scala)
/tmp/sbt_3bc84e4d/xsbt/Compat.scala:152: error: postfix operator headOption needs to be enabled
by making the implicit value scala.language.postfixOps visible.
This can be achieved by adding the import clause 'import scala.language.postfixOps'
or by setting the compiler option -language:postfixOps.
See the Scaladoc for value scala.language.postfixOps for a discussion
why the feature needs to be explicitly enabled.
          } headOption
            ^
one error found
```

**Comment by som-snytt** [comment]
The compat kludge is improved in https://github.com/scala/scala/pull/7319

**Comment by som-snytt** [comment]
There ought to be a scalafix for:
```
for (e <- it(coll)) f(e) foreach (b +=)
```

**Comment by som-snytt** [comment]
That "double-feature" test has a clever name, but it relies on postfix to trigger, probably because typechecking happens twice but not silently. But it won't be reported twice as an error. Ironic! as Alanis would say. The test should be changed to just report twice manually.

**Comment by lrytz** [comment]
This seems to be ready. @SethTisue merge?

**Comment by SethTisue** [comment]
I need to brace myself for the likely avalanche of needed project tweaks in the community build first :-)

**Comment by lrytz** [comment]
Ah, right :-)

**Comment by SethTisue** [comment]
@som-snytt would you mind updating the PR description to briefly describe and justify the change, since it will be linked from the release notes?I think this change could be surprising and annoying for some users, so I think we should have some justification only one click away from the release notes. (I'm strongly in favor of the change regardless, of course. language simplification, onward!)

**Comment by SethTisue** [comment]
@odersky when you're listing language features that are being removed, e.g. procedure syntax, you might like to include this as well

**Comment by SethTisue** [comment]
> I need to brace myself for the likely avalanche of needed project tweaks in the community build first

it has been a puny avalanche so far, only a few pebbles here and there 👍

**Pull request merged**